apps/prairielearn/src/ee/pages/lti13CourseNavigation/lti13CourseNavigation.ts

```diff
@@ -1,10 +1,192 @@
+        context_id: ltiClaim.context?.id,
+      },
+      Lti13CourseInstanceSchema,
+    );
+
+    if (lti13_course_instance) {
+      // Update lti13_course_instance on instructor login
+      // helpful as LMS updates or we add features
+      if (role_instructor) {
+        await queryAsync(sql.update_lti13_course_instance, {
+          lti13_instance_id: req.params.lti13_instance_id,
+          course_instance_id: lti13_course_instance.course_instance_id,
+          deployment_id: ltiClaim.deployment_id,
+          context_id: ltiClaim.context?.id,
+          context_label: ltiClaim.context?.label,
+          context_title: ltiClaim.context?.title,
+        });
+
+        // TODO: Set course/instance staff permissions for LMS course staff here?
+      }
+
+      // LTI claims are not used after this page so remove them from the session
+      ltiClaim.remove();
+
+      // Redirect to linked course instance
+      res.redirect(
+        `/pl/course_instance/${lti13_course_instance.course_instance_id}/${
+          role_instructor ? 'instructor/' : ''
+        }`,
+      );
+      return;
+    }
+
+    // Students get a "come back later" message
+    if (!role_instructor) {
+      res.send(
+        Lti13CourseNavigationNotReady({
+          resLocals: res.locals,
+          courseName,
+        }),
+      );
+      return;
+    }
+
+    // Instructor so lookup their existing information in PL
+    const courses = await selectCoursesWithEditAccess({
+      user_id: res.locals.authn_user.user_id,
+      is_administrator: res.locals.authn_is_administrator,
+    });
+
+    const course_instances: CourseInstanceAuthz[] = [];
+
+    for (const course of courses) {
+      // Only course owners can link
+      if (!course.permissions_course.has_course_permission_own) {
+        continue;
+      }
+
+      const instances = await selectCourseInstancesWithStaffAccess({
+        course_id: course.id,
+        user_id: res.locals.authn_user.user_id,
+        authn_user_id: res.locals.authn_user.user_id,
+        is_administrator: res.locals.authn_is_administrator,
+        authn_is_administrator: res.locals.authn_is_administrator,
+      });
+
+      course_instances.push(...instances);
+    }
+
+    res.send(
+      Lti13CourseNavigationInstructor({
+        resLocals: res.locals,
+        courseName,
+        courses,
+        course_instances,
+      }),
+    );
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    const unsafe_course_instance_id = req.body.unsafe_course_instance_id;
+
+    const ltiClaim = new Lti13Claim(req);
+    const authn_lti13_instance_id = req.session.authn_lti13_instance_id;
+
+    // Map passed and auth lti13_instance_id through institution to course instance, or fail
+    await queryOneRowAsync(sql.select_lti13_course_instance_institution, {
+      course_instance_id: unsafe_course_instance_id,
+      lti13_instance_id: req.params.lti13_instance_id,
+      authn_lti13_instance_id,
+    });
```

Comment:
Nit: as the person who gets paged on 500s, I prefer to jump through hoops to translate these failures into non-5XX status codes. In this case, I'd do a `queryOptionalRow` followed by an `if == null throw new HttpStatusError(...)`. A 403 is probably appropriate here? We can think about that a bit.

Reply:
Addressed in 10bfa667d